intelmq/bots/parsers/key_value/parser.py

```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+"""
+Parse a string of key=value pairs.
+
+SPDX-FileCopyrightText: 2020 Linköping University <https://liu.se/>
+SPDX-License-Identifier: AGPL-3.0-or-later
+
+Tokens which do not contain the kv_separator string are ignored.
+
+Values cannot contain newlines.
+
+Parameters:
+
+    pair_separator: string, default ' ', string separating key=value
+                    pairs
+
+    kv_separator: string, default '=', string separating key and value
+
+    keys: array of strings to strings, names of keys -> names of
+          fields to propagate
+
+    strip_quotes: boolean, default true, remove opening and closing
+                  quotes
+
+    timestamp_key: string, optional, key containing event timestamp.
+                   Numerical values are interpreted as UNIX seconds,
+                   others are parsed by
+                   dateutil.parser.parse(fuzzy=True). If parsing fails
+                   no timestamp field will be added.
+
+"""
+
+from intelmq.lib import utils
+from intelmq.lib.bot import ParserBot
+from intelmq.lib.exceptions import ConfigurationError
+from intelmq.lib.harmonization import DateTime
+
+from dateutil.parser import parse
+
+
+class KeyValueParserBot(ParserBot):
+
+    def init(self):
+        self.pair_separator = getattr(self.parameters, 'pair_separator', ' ')
+        self.kv_separator = getattr(self.parameters, 'kv_separator', '=')
+        self.keys = getattr(self.parameters, 'keys', {})
+        if not self.keys:
+            raise ConfigurationError('Key extraction', 'No keys specified.')
+        self.strip_quotes = getattr(self.parameters, "strip_quotes", True)
+        self.timestamp_key = getattr(self.parameters, "timestamp_key", None)
+        if self.timestamp_key:
+            self.keys[self.timestamp_key] = "time.source"
+
+    def parse(self, report):
+        raw_report = utils.base64_decode(report.get('raw'))
+        for row in raw_report.splitlines():
+            yield row
+
+    def parse_line(self, row, report):
+        event = self.new_event(report)
+        for kv_pair in row.split(self.pair_separator):
+            (key, _, value) = kv_pair.rpartition(self.kv_separator)
+            if not key:
+                continue
+            if self.strip_quotes and value.startswith('"') and value.endswith('"'):
+                value = value[1:-1]
+            if key == self.timestamp_key:
+                try:
+                    if value.isnumeric():
+                        value = DateTime.from_timestamp(int(value))
+                    else:
+                        value = parse(value, fuzzy=True).isoformat() + " UTC"
+                except:
+                    value = None  # Will be ignored by event.add()
```

Comment:
If we can't parse it, do we *really* want to ignore this? That drops data.

Reply:
This is correct. However, it only drops the timestamp, and lets the rest of the fields through. A garbage timestamp will fail harmonisation and lead to the entire event being dropped.

I think. But apparently my development environment is halfway between working states at the moment, so I haven't actually tried.

Reply:
I have now repaired my development environment. What happens when trying to add an unparseable timestamp field is that if `event.add()` is called with `raise_failure=False`, the field is silently discarded. If `raise_failure=True` (which is the default), the whole message is discarded and an error is noted in the bot's log file. So I think that throwing away the timestamp is better than throwing away the entire message.

Reply:
I now added a warning log message in case this happens.